cachi2/core/package_managers/npm.py

```diff
@@ -502,6 +566,95 @@ def _get_npm_dependencies(
     return download_paths
 
 
+def replace_dependency(dependency_version: str) -> bool:
+    """Check if dependency must be updated in package(-lock).json.
+
+    package(-lock).json files require to replace dependency URLs for
+    empty string in git and https dependencies in V2+.
+    """
+    url = urlparse(dependency_version)
+    if url.scheme == "file":
+        return False
+    return not (url.scheme == "" and "/" not in dependency_version)
+
+
+def update_package_lock_with_local_paths(
+    download_paths: Dict[NormalizedUrl, RootedPath],
+    package_lock: PackageLock,
+) -> None:
+    """Replace packages resolved URLs with local paths.
+
+    Update package-lock.json file in a way it can be used in isolated environment (container)
+    without internet connection. All package resolved URLs will be replaced for
+    local paths to downloaded dependencies.
+
+    :param download_paths:
+    :param package_lock: PackageLock instance which holds package-lock.json content
+    """
+    for package in package_lock.packages:
+        # Remove integrity for git sources, their integrity checksum will change when
+        # constructing tar archive from cloned repository
+        if package.resolved_url:
+            url = _normalize_resolved_url(str(package.resolved_url))
+        else:
+            continue
+        if _classify_resolved_url(url) == "git":
+            if package.integrity:
+                package.integrity = ""
+
+        # Update all type of dependencies except file:
+        if _classify_resolved_url(url) != "file":
+            templated_abspath = Path("${output_dir}", download_paths[url].subpath_from_root)
+            package.resolved_url = f"file://{templated_abspath}"
+
+        for dep_type in DEPENDENCY_TYPES:
+            if package.package_dict.get(dep_type):
+                for dependency, dependency_version in package.package_dict[dep_type].items():
+                    if replace_dependency(dependency_version):
+                        package.package_dict[dep_type].update({dependency: ""})
+
+    main_package = package_lock.lockfile_data["packages"][""]
+    for dep_type in DEPENDENCY_TYPES:
+        if main_package.get(dep_type):
+            for dependency, dependency_version in main_package[dep_type].items():
+                if replace_dependency(dependency_version):
+                    main_package[dep_type].update({dependency: ""})
```

Comment:
Suggestion: If the `get_main_package` function returned a `Package`, we could rewrite the for loop here as 

```
for package in package_lock.packages + [package_lock.main_package]:
```
And this block here could be completely removed.

Reply:
Or maybe we could treat the main package as part of `PackageLock._packages`, and only separate it in the end of the `_resolve_npm` function.

Reply:
I'll keep lockfile_data as it is, but will try to package_lock.main_package property which will return Instance of Package class

Reply:
Yeah, I agree with @chmeliik. Problem is, we need the original data to do the replacements. Maybe we could return a `deepcopy` of the dictionary from the `lockfile_data` [function](https://github.com/containerbuildsystem/cachi2/blob/c4d277d0d9f7bbca93c6ecb8abf1152a3857ac50/cachi2/core/package_managers/npm.py#L179)?

Should be easy enough, and that way we'd still keep the encapsulation.

Reply:
Well, the Package class also exposes the `_package_dict`, which then gets modified externally. We can improve this later, right? :lolsob:

Reply:
I'll try to polish it in different PR. ;)